### ISSUE TITLE
Update schedule to signal converter handling of phase accumulation

### ISSUE THREAD

**Issue by DanPuzzuoli** [Qiskit/qiskit-dynamics#140]
The schedule -> signal converter currently ensures that the output signals satisfy Equation (1) of [this paper](https://iopscience.iop.org/article/10.1088/2058-9565/aba404/meta), with the frequency interpreted as the current channel frequency value given by `SetFrequency`/`ShiftFrequency` instructions, and the phase interpreted as being the value given purely by `SetPhase`/`ShiftPhase` instructions.

As pointed out by @nkanazawa1989 in [this notebook](https://gist.github.com/nkanazawa1989/460909fd56e326d7796d813fbeaed089) however, frequency changes result in phase accumulation on real IBM backends. Hence, the phase in Equation (1) should actually be a function of the `SetPhase`/`ShiftPhase` instructions, _as well as_ the `SetFrequency`/`ShiftFrequency` instructions (and the times at which they occur).

We should update the behaviour of the pulse -> signal converter to be consistent with this.

**Comment by DanPuzzuoli** [comment]
I guess the mathematical condition is to keep `exp(i(2 * pi * f * t + phi))` continuous under frequency changes? So if `f` is shifted `phi` needs to be shifted to preserve continuity?

Is it enough to change [lines 137-138](https://github.com/Qiskit/qiskit-dynamics/blob/f8a34596285d0ff3fb5d701f05311e6919f1a809/qiskit_dynamics/pulse/pulse_to_signals.py#L137-L138) to also shift the phase (and similarly [lines 143-144](https://github.com/Qiskit/qiskit-dynamics/blob/f8a34596285d0ff3fb5d701f05311e6919f1a809/qiskit_dynamics/pulse/pulse_to_signals.py#L143-L144) for the `SetFrequency` instruction?).

E.g. something like:
```
    if isinstance(inst, ShiftFrequency):
        phases[chan] += (frequency_shifts[chan] - inst.frequency) * start_time * self._dt
        frequency_shifts[chan] += inst.frequency

```
Not sure I have the exact correct time or signs, but I think something along these lines should work.

**Comment by to24toro** [comment]
Thank you for your response.
Exaclty, your idea is much similar to my first idea. The only difference is whether the phase accumulation term is explicit or embedded in the phase. Of course, I think that this is easier to implement.
Only my concern is that the adding term to phase becomes ssomewhat complex. However the implementation is very simple, so if you don't mind, I want to choose this method.

And let me add one more thing. I think that t is t is different from start_time. It is necessary to keep the time of shifting frequency because you need to accumulate for  $\Delta$ t.

![IMG_0117](https://user-images.githubusercontent.com/38037695/202593820-65dbb49a-ea7f-4d00-9d97-aa7ace0d7261.jpg)

**Comment by DanPuzzuoli** [comment]
Hmm interesting - can you write down what the mathematical formula should be in this case?

It's a good point though that it should maybe be tracked separately from the phase, as if phase is accumulated due to `ShiftFrequency`/`SetFrequency`, and we store it in the phase, this information could be accidentally erased by a `SetPhase` instruction. This is maybe another thing to check: what is the expected interaction between `ShiftFrequency`/`SetFrequency` and `SetPhase`.

**Comment by to24toro** [comment]
The below code doesn't represent a methematical fomula but you maybe understand the phase accumulation if you read it.
The point is that 2 * pi * $\Delta$ f *  $\Delta$ t is phase-accumalated when shifting frequency.
```
import numpy as np
import matplotlib.pyplot as plt
import random
X = np.linspace(0,3,601)

# prepare for shift frequency times and two frequencies
sf1 = random.randint(101,200)/100
sf2 = random.randint(201,300)/100
f1 = 1
f2 = 0.7
x_f1 = []
x_f2 = []
y_f1 = []
y_f2 = []
for x in X:
    if x < sf1:
        x_f1.append(x)
        y_f1.append(np.sin(2 * np.pi * f1 * x))
    elif x < sf2:
        x_f2.append(x)
        # phase accumulation 2 * pi * (f1-f2)*(sf1 -0)
        y_f2.append(np.sin(2 * np.pi * (f2 * x + (f1-f2)*sf1)))
    else:
        x_f1.append(x)
        # phase accumulation 2 * pi * (f2-f1)*(sf2 -sf1)
        y_f1.append(np.sin(2 * np.pi * (f1 * x + (f2-f1)*(sf2-sf1))))

plt.scatter(x_f1,y_f1,color = "red")
plt.scatter(x_f2,y_f2,color = "blue")
```
![output](https://user-images.githubusercontent.com/38037695/202889045-9ee737da-91ec-4cd0-8654-fef89662ec6f.png)

**Comment by DanPuzzuoli** [comment]
Thank you, I think I understand. Can you slightly modify your example so that the second `SetFrequency` is to a third frequency `f3`, just to make it fully general?

So just to confirm what you've said: The phase accumulation due to `SetFrequency`/`ShiftFrequency`, call it `freq_phase`, needs to be tracked _separately_ from the phase being set by `SetPhase`/`ShiftPhase`, as otherwise `SetPhase` could accidentally erase the phase accumulation. Is this correct?

That seems reasonable, but I think we may need to create an example, similar to @nkanazawa1989 's original example that includes `SetFrequency`/`ShiftFrequency` instructions, AND a `SetPhase` instruction, to confirm how real backends interpret combinations of these instructions. What do you think?

**Comment by nkanazawa1989** [comment]
ASAIK the `SetPhase` is not supported by IBM hardware (only `ShiftPhase` is available). 
![image](https://user-images.githubusercontent.com/39517270/202946209-a4bcecda-12ed-4482-8a50-c99c8dc09e48.png)
But I agree we need some test case that include both commands. I guess shift phase command will introduce phase discontinuity, but at this point it is natural to assume amp=0 because we cannot interleave the frame moderation command during the pulse and pulse usually starts and ends with amp=0. Otherwise, such discontinuity cannot be perfectly reproduced on hardware because of the transient phenomenon, and it could also cause issue in simulator because the master equation might diverge.

**Comment by nkanazawa1989** [comment]
The current frame in the dynamics is semi-stateful because it only knows the current phase and frequency value. To consider the phase accumulation it must need to know the time at which the frame was previously modulated. 

> Thank you, I think I understand. Can you slightly modify your example so that the second SetFrequency is to a third frequency f3, just to make it fully general?

This is good question. We would naively reference to f2, but I'm not sure if this is really expected behavior.

**Comment by to24toro** [comment]
I updated [my example](https://github.com/Qiskit/qiskit-dynamics/issues/140#issuecomment-1321051038).
As a code, it is enough to add $\Delta$ f *  $\Delta$ t to freq_phases[chan] every time SetFrequency/ShiftFrequency.

**Comment by DanPuzzuoli** [comment]
> I updated [my example](https://github.com/Qiskit/qiskit-dynamics/issues/140#issuecomment-1321051038). As a code, it is enough to add Δ f * Δ t to freq_phases[chan] every time SetFrequency/ShiftFrequency.

Is it `Δ f * Δ t` or just `Δ f * t`, where `t` is the time at which the shift is happening? Your code above seems to be this latter one. Either way seems like you've got it figured out now!

**Comment by to24toro** [comment]
Sorry for confusion. As a result, it is Ok to use $\Delta$ f *  t , which corresponds to `start_time * self._dt`  you showed.
It will be resolved among us. I would like to show the demo about this in order to confirm whether this behavior is consistent with that of readl device  in the pulse meeting (https://ibm-quantumcomputing.slack.com/archives/GFHBSC3EY/p1668997127965569).

**Comment by DanPuzzuoli** [comment]
Closed by #164
